File: catalog/backend/src/Models/Variant.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Runtime.Serialization;
using System.Text.Json.Serialization;

namespace backend.Models;


public class VariantDTO
{
    public static string defaultPrePackedItem = "No";

    [Required]
    [ForeignKey("Product")]
    public int? ProductId { get; set; }

    public string? PrePackedItem { get; set; } = defaultPrePackedItem;

    [Required]
    public string? SKU { get; set; }

    [Required]
    public string? Barcode { get; set; }

    [Required]
    public string VariantDescription { get; set; } = "";

    [Required]
    public string Composition { get; set; } = "";

    [Required]
    public string? Hscode { get; set; }

    [Required]
    public string? Country { get; set; }

    public string? Gender { get; set; }

    public string? AgeGroup { get; set; }

    public string? Color { get; set; }

    public string? Size { get; set; }

    public string? Fabric { get; set; }

    [Required]
    public double? Width { get; set; }

    [Required]
    public double? Height { get; set; }

    [Required]
    public double? Depth { get; set; }

    [Required]
    public double? Weight { get; set; }

    public double? PriceRetail { get; set; }

    public double? PriceWholesale { get; set; }

    public double? Cost { get; set; }

    [Required]
    public string? Currency { get; set; }

    public int? Dangerous { get; set; }

    public int? UN { get; set; }

    public int? PackingCode { get; set; }

    public bool IsArchived { get; set; } = false;
}


public class Variant
{
    static public Variant VariantWithId(int id, VariantDTO variantDto)
    {
        Variant variant = new Variant(variantDto);
        variant.Id = id;
        return variant;
    }

    public Variant()
    {
    }

    public Variant(VariantDTO variantDTO) : base()
    {
        ProductId = variantDTO.ProductId;
        PrePackedItem = variantDTO.PrePackedItem;
        SKU = variantDTO.SKU;
        Barcode = variantDTO.Barcode;
        VariantDescription = variantDTO.VariantDescription;
        Composition = variantDTO.Composition;
        Hscode = variantDTO.Hscode;
        Country = variantDTO.Country;
        Gender = variantDTO.Gender;
        AgeGroup = variantDTO.AgeGroup;
        Color = variantDTO.Color;
        Size = variantDTO.Size;
        Fabric = variantDTO.Fabric;
        Width = variantDTO.Width;
        Height = variantDTO.Height;
        Depth = variantDTO.Depth;
        Weight = variantDTO.Weight;
        PriceRetail = variantDTO.PriceRetail;
        PriceWholesale = variantDTO.PriceWholesale;
        Cost = variantDTO.Cost;
        Currency = variantDTO.Currency;
        Dangerous = variantDTO.Dangerous;
        UN = variantDTO.UN;
        PackingCode = variantDTO.PackingCode;
        IsArchived = variantDTO.IsArchived;
    }

    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }

    [Required]
    [ForeignKey("Product")]
    public int? ProductId { get; set; }

    [Required]
    [JsonIgnore]
    public Product Product { get; set; } = null!;

    public string? PrePackedItem { get; set; } = VariantDTO.defaultPrePackedItem;

    [Required]
    public string? SKU { get; set; }

    [Required]
    public string? Barcode { get; set; }

    [Required]
    public string VariantDescription { get; set; } = "";

    [Required]
    public string Composition { get; set; } = "";

    [Required]
    public string? Hscode { get; set; }

    // Iso 3166-1 Alpha 2 (iso2code)
    [Required]
    public string? Country { get; set; }

    public string? Gender { get; set; }

    public string? AgeGroup { get; set; }

    public string? Color { get; set; }

    public string? Size { get; set; }

    public string? Fabric { get; set; }

    [Required]
    public double? Width { get; set; }

    [Required]
    public double? Height { get; set; }

    [Required]
    public double? Depth { get; set; }

    [Required]
    public double? Weight { get; set; }

    public double? PriceRetail { get; set; }

    public double? PriceWholesale { get; set; }

    public double? Cost { get; set; }

    // Iso 4217 (iso3code)
    [Required]
    public string? Currency { get; set; }

    public int? Dangerous { get; set; }

    public int? UN { get; set; }

    public int? PackingCode { get; set; }

    public bool IsArchived { get; set; } = false;
}

public class IsArchivedDTO
{
    public bool IsArchived { get; set; }
}

File: catalog/frontend/lib/model/variant.dart
import 'package:frontend/model/product.dart';

class Variant {
  final Product product;
  final num? id;
  String barcode;
  String sku;
  String? prePackedItem;
  String variantDescription;
  String composition;
  String hscode;
  String country;
  String? gender;
  String? ageGroup;
  String? color;
  String? size;
  String? fabric;
  num width;
  num height;
  num depth;
  num weight;
  num? priceRetail;
  num? priceWholesale;
  num? cost;
  String currency;
  num? dangerous;
  num? unNumber;
  num? packingCode;
  bool isArchived = false;

  Variant(
      {required this.product,
      this.id,
      required this.sku,
      required this.barcode,
      this.prePackedItem,
      required this.variantDescription,
      required this.composition,
      required this.hscode,
      required this.country,
      this.gender,
      this.ageGroup,
      this.color,
      this.size,
      this.fabric,
      required this.width,
      required this.height,
      required this.depth,
      required this.weight,
      this.priceRetail,
      this.priceWholesale,
      this.cost,
      required this.currency,
      this.dangerous,
      this.unNumber,
      this.packingCode,
      required this.isArchived});

  Map<String, dynamic> toJson() => {
        'id': id,
        'productId': product.id,
        'barcode': barcode,
        'sku': sku,
        'prePackedItem': prePackedItem,
        'variantDescription': variantDescription,
        'composition': composition,
        'hscode': hscode,
        'country': country,
        'gender': gender,
        'ageGroup': ageGroup,
        'color': color,
        'size': size,
        'fabric': fabric,
        'width': width,
        'height': height,
        'depth': depth,
        'weight': weight,
        'priceRetail': priceRetail,
        'priceWholesale': priceWholesale,
        'cost': cost,
        'currency': currency,
        'dangerous': dangerous,
        'un': unNumber,
        'packingCode': packingCode,
        'isArchived': isArchived
      };

  static Variant fromJson(Map<String, dynamic> json, Product product) {
    return Variant(
        product: product,
        id: json['id'],
        sku: json['sku'],
        barcode: json['barcode'],
        variantDescription: json['variantDescription'],
        composition: json['composition'],
        hscode: json['hscode'],
        country: json['country'],
        gender: json['gender'],
        ageGroup: json['ageGroup'],
        color: json['color'],
        size: json['size'],
        fabric: json['fabric'],
        width: json['width'],
        height: json['height'],
        depth: json['depth'],
        weight: json['weight'],
        priceRetail: json['priceRetail'],
        priceWholesale: json['priceWholesale'],
        cost: json['cost'],
        currency: json['currency'],
        dangerous: json['dangerous'],
        unNumber: json['un'],
        packingCode: json['packingCode'],
        prePackedItem: json['prePackedItem'],
        isArchived: json['isArchived']);
  }
}
